pymeasure/instruments/anaheimautomation/stepmotorcontroller.py

```diff
@@ -0,0 +1,251 @@
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+#
+
+import asyncio
+from pyvisa.errors import VisaIOError
+from pymeasure.adapters import VISAAdapter
+from pymeasure.instruments import Instrument
+from pymeasure.instruments.validators import truncated_range, strict_discrete_set
+
+
+class MotorController(Instrument):
+    """Driver to interface with Anaheim Automation stepper motor controllers. This driver has been tested with the DPY50601 and DPE25601 motor controllers. 
+    """
+
+    basespeed = Instrument.control(
+        "VB", "B%i", 
+        """Integer property that represents the motor controller's starting/homing speed. This property can be set.""",
+        validator=truncated_range,
+        values=[1, 5000],
+    )
+    
+    maxspeed = Instrument.control(
+        "VM", "M%i", 
+        """Integer property that represents the motor controller's maximum (running) speed. This property can be set.""",
+        validator=truncated_range,
+        values=[1, 50000],
+    )
+
+    direction = Instrument.control(
+        "V+", "%s",
+        """A string property that represents the direction in which the stepper motor will rotate upon subsequent step commands. This property can be set. 'CW' corresponds to clockwise rotation and 'CCW' corresponds to counter-clockwise rotation.""",
+        map_values=True, 
+        validator=strict_discrete_set,
+        values={"CW" : "+", "CCW" : "-"},
+        get_process=lambda d: "+" if d == 1.0 else "-",
+    )    
+
+    steps = Instrument.control(
+        "VN", "N%i",
+        """An integer property that represents the number of steps the motor controller will run upon the next 'G' (go) command that is sent to the controller. This property can be set.""",
+        validator=truncated_range,
+        values=[0, 8388607],
+    )
+
+    position = Instrument.control(
+       "VZ", "Z%i",
+       """An integer property that represents the step position as seen by the motor controller. This property can be set.""",
+       validator=truncated_range,
+       values=[-8388607, 8388607],
+    ) 
+
+    encoder_position = Instrument.measurement(
+        "VEP", 
+        """An integer property that represents the step position as counted by an externally connected encoder. This property cannot be set, but can be 
+        reset to 0 using the reset_encoder_position() method.""",
+    )
```

Comment:
I like this suggestion, though one issue I see with merging these two properties is that the `motor position` can be set to any value within the range that I have specified in the current `position` property, while the `encoder position` can only be reset to 0. I think mixing these properties could make this fact less transparent. What are your thoughts on only allowing the position to be reset to 0? So there would be a single `position` property to query either the motor or encoder position as you have described, then a separate method `reset_position()` is created that always resets both the motor position and the encoder position.

Reply:
For my use case, I will never need to be able to set the `motor position` to an arbitrary value. I will only ever need to reset it to 0 when performing homing operations. My suspicion is that this will be the case for the vast majority of users as well.

Reply:
>  So there would be a single position property to query either the motor or encoder position as you have described, then a separate method reset_position() is created that always resets both the motor position and the encoder position.

Sounds good. Maybe call it `home()` or something similar if that matches what you do.

> For my use case, I will never need to be able to set the motor position to an arbitrary value. I will only ever need to reset it to 0 when performing homing operations. My suspicion is that this will be the case for the vast majority of users as well.

Hm, it seems I have misunderstood something. I was under the impression that you can also command the motor to move by setting a position, i.e. `motor.position = 1234` moves to that point (e.g. by using  `step()`in the background. If that is not the case, why? Would that not be a quite natural interface?

Reply:
I can certainly move the driver in that direction. Originally, the `position` property was only used to get/set the value of the controller's internal position reference. It wouldn't actually move the motor at all. Though I think being able to do `motor.position = 1234` and have the motor move to that position would be quite nice. I'd also like to distinguish between _step position_ and _absolute position_. The motor controller always counts position measured in steps, though in my use case I'd like to be able to command an _absolute_ position in say meters, inches, etc. 

A potential way this can be accomplished is to have two properties, one called `step_position` which behaves exactly like the current `position` property but with the added feature that you can run `motor.step_position = <desired position>` and the motor will actually move to that position. The other property would be called `absolute_position` and it also behaves exactly the same way, but calls unit conversion methods `absolute_to_steps()` when setting a new position and `steps_to_absolute()` when querying position. These methods would of course have to be implemented in a subclass as the mapping between steps and absolute position will vary depending on what the user is driving with the controller. What do you think of this type of interface? My next commit will include my idea for this implementation.

Reply:
> Sounds good. Maybe call it `home()` or something similar if that matches what you do.
The problem with calling it home() is that homing is a command that can be sent to the controller, so I'd like to have a separate home() method to perform this operation. I've just implemented this method and kept the name `reset_position()` for the other method.